### PR TITLE
[E1-7][P2] Issue テンプレートの整備（ループが読み取れる形式に揃える）

### LINKED ISSUE #37: [E1-7][P2] Issue テンプレートの整備（ループが読み取れる形式に揃える）
## 目的

Issue の書式を固定し、ループ（自動実行）が本文から必要な情報を**推測せずに**読み取れるようにする。

Issue #1 / #2 / #3 / #5 / #6 / #9 を自動で実装した5周で、**5本の PR のうち4本に「判断を仰ぎたい点」を書く**ことになった。その多くは実装の難しさではなく、Issue から読み取れなかった情報を推測で埋めたことに起因する。テンプレートでこの往復を前倒しする。

### 1. 依存がエピック ID で書かれており、Issue 番号への解決が壊れやすい

各 Issue の `## 依存` は `E2-1` のようなエピック ID で書かれている。これを Issue 番号に解決するには、全 Issue のタイトル接頭辞 `[E2-1]` と突き合わせる必要があり、`.claude/skills/loop-run/SKILL.md` にもその手順が明記されている。

**接頭辞が1文字違っていた場合、依存が無いものとして静かに着手する。** 検知する手段がない。

### 2. スコープ外が書かれていないため、境界を他の Issue から推測した

以下はすべて、担当 Issue の本文ではなく**他の Issue の本文を読んで**「自分の担当外」と判断した。

| 推測した境界 | 根拠にした Issue |
| --- | --- |
| 時間の丸めはやらない | #7 |
| タイムゾーンの扱いはやらない | #22 |
| タグの正規化・階層パースはやらない | #8 |
| 引数パース・ヘルプ・終了コードはやらない | #12 |
| 日跨ぎ分割・重複判定はやらない（#5 で） | #6 |

今回は結果的に当たっていたが、AI が最も間違えやすい種類の推測である。

### 3. DoD と目的が食い違い、1往復増えた

#2 は DoD が「`check` が **typecheck・lint・test** をすべて実行し」と3つを列挙する一方、目的には「1コマンドで**全部**チェック」とあった。整形チェックを含めるべきか判断できず、DoD の文面を優先して実装し、PR 本文で判断を預けた。その後レビューで「`format:check` も入れる」と方針が確定し、修正した（PR #31）。

DoD が唯一の完了定義であることは CLAUDE.md に書かれているが、**目的の側に要件が書かれている**とこの食い違いが起きる。

### 4. 技術選定の希望が本文ではなくコメントで来た

#2 の「oxlint,oxfmt でもいいかも」は Issue 本文ではなくコメントだった。停止コメントの確認のためにコメントを読む手順があったので拾えたが、**着手後に投稿されていたら見落とす。**

### 5. 両義的な指定を判断できず、両方実装した

#9 の「保存先はテストで差し替え可能にする（環境変数**または**引数）」。どちらか決められず両方用意した。

## スコープ

- `.github/ISSUE_TEMPLATE/` にバックログ項目用のテンプレートを追加する
- **Issue Forms（`.yml`）を使う。** フィールドが固定の見出しとして本文に出るため、機械的に読む側が位置を推測しなくてよい
- 用意するフィールド（案）

| フィールド | 目的 |
| --- | --- |
| 目的 | 何のためか。**要件は書かない** |
| スコープ | やること |
| **スコープ外** | やらないこと。他 Issue に投げる場合は番号を書く |
| 依存 | **Issue 番号**（`#5`）で書く。エピック ID を使わない |
| 受け入れ条件 (DoD) | チェックリスト。**これが唯一の完了定義**である旨を説明文に明記 |
| 技術選定の希望・制約 | ライブラリやツールの指定。空欄なら実装側の判断でよい |
| テストで固定すべき境界値 | 0件・同時刻・日跨ぎ・空文字・上限値など、Issue 側から指定したいもの |

- テンプレート導入後、`.claude/skills/loop-run/SKILL.md` の「2-1 候補の抽出」からエピック ID の突き合わせ手順を削れるか検討する

## 受け入れ条件 (DoD)

- [ ] `.github/ISSUE_TEMPLATE/` にバックログ項目用のテンプレートがある
- [ ] テンプレートに「スコープ外」「技術選定の希望・制約」「テストで固定すべき境界値」のフィールドがある
- [ ] 依存を Issue 番号で書く形式になっており、説明文でエピック ID を使わないよう案内している
- [ ] DoD が唯一の完了定義であることが、テンプレートの説明文に書かれている
- [ ] テンプレートから実際に Issue を1件作成し、`loop-run` が依存とスコープ外を推測せずに読めることを確認した

## 検討が必要な点（この Issue の判断に含めたい）

- **既存 Issue の移行をどうするか。** 現在 open な26件は依存がエピック ID で書かれている。全件を Issue 番号に書き換えるのか、新規分だけテンプレートに従うのかを決める必要がある。書き換えるなら別 Issue に分けたほうがよい
- **GitHub のサブイシュー機能を使うか。** `## 依存` の代わりにネイティブの親子関係で表現する選択肢がある。ただし `loop-run` 側の読み取り方法を変える必要がある

## 補足：テンプレートで防げないもの

この5周でレビューが検出した実バグは2件あった。

- `2026-08-12T24:00:00Z` が翌日として保存される（PR #33）
- 実行中エントリが範囲列挙から落ち、前夜から続く作業が当日の集計から消える（PR #35）

**どちらも仕様の曖昧さではなく、実装側のテストの穴が原因である。** テンプレートは「人に確認する往復の数」に効くが、「バグの数」には効かない。後者はレビューとテストの側の課題として切り分けておく。

## 依存

なし（#4 で CLAUDE.md が整備されるが、テンプレート自体は独立して追加できる）

### PR DESCRIPTION
**Pull request by okrago10**
## 概要

`.github/ISSUE_TEMPLATE/backlog-item.yml` を追加し、テンプレートが方針から外れたら
`npm run check` が落ちるよう `tests/github/` に検査を置いた。

Closes #37

### 設計の理由

**自由記述の Markdown ではなく Issue Forms（`.yml`）を使う。** Issue の指定どおり。
フィールドが固定の見出しとして本文に出るため、読む側が位置を推測しなくてよい。
GitHub は各 textarea を `### <label>` として本文に展開するので、`loop-run` は
見出しを目印に引ける（#70 の本文で実際に確認した。下記 DoD の最後の項目）。

**「スコープ外」を必須にした。** #37 の背景2 のとおり、これまで実装側は担当 Issue の
本文ではなく**他の Issue の本文を読んで**境界を推測していた（時間の丸め → #7、
タイムゾーン → #22 など5件）。必須にして「なければ『なし』と書く」ことにすれば、
空欄が「未記入」か「本当に無い」かを区別できる。

**「技術選定の希望・制約」は任意にし、空欄の意味を説明文に書いた。** #2 の
「oxlint,oxfmt でもいいかも」が本文ではなくコメントで来た件（背景4）への対応。
欄があれば本文に書かれる。空欄は「実装側の判断でよい」と読める。

**依存は Issue 番号で書かせる。** 背景1 のとおり、エピックIDだけだとタイトル接頭辞
との突き合わせが必要で、接頭辞が1文字違うと**依存が無いものとして静かに着手する**。
`E2-1（#5）` の併記も認めるが、番号は必須。

**`config.yml` で自由記述の Issue を残した。** バックログ項目は揃えたいが、不具合の
速報・運用の相談・ふりかえりのメモまで同じ枠に押し込むと、枠に収まらないときに
Issue が作られなくなる。テンプレートは「選べる形」として置く。

**YAML の完全な解析はしない。** 検査は Issue Forms が実際に使う範囲（トップレベルの
キー、`body` の並び、`id` / `attributes` / `validations`、ブロックスカラー）だけを読む。
`CLAUDE.md`「依存ライブラリの追加ルール」に従い、パーサを入れずに済ませた
（`tests/workflows/action-pinning.ts` が `uses:` を行単位で見ているのと同じ方針）。
**解析できない書き方が「違反0件」で通らないよう、`type` が読めない項目は違反として
報告する。** これは action-pinning で実際に踏んだ穴と同じ形の対策。

なお**部分解析が実 YAML と一致していることは別途確認した**（下記「検証」）。

## 受け入れ条件

- [x] **`.github/ISSUE_TEMPLATE/` にバックログ項目用のテンプレートがある**
      → `tests/github/issue-template.test.ts` の「バックログ項目用のテンプレートが1本以上ある」

      実際に読み取れたフィールド:
      ```
      name: バックログ項目
      title: '[E?-?][P?] '

      id                     req      label
      ------------------------------------------------------------
      (なし)                   -        (冒頭の案内)
      purpose                True     目的
      scope                  True     スコープ
      out-of-scope           True     スコープ外
      dependencies           True     依存
      acceptance-criteria    True     受け入れ条件 (DoD)
      tech-constraints       False    技術選定の希望・制約
      boundary-values        False    テストで固定すべき境界値
      ```

- [x] **テンプレートに「スコープ外」「技術選定の希望・制約」「テストで固定すべき境界値」の
      フィールドがある**
      → `tests/github/issue-template.test.ts`「`out-of-scope`（スコープ外）のフィールドがある」
      ほか、`REQUIRED_FIELDS` の7件それぞれに1テスト。上の表の `id` 列がその出力。

      **`id` と見出しの両方を固定している。** `id` だけを見ていると見出しの文言が変わった
      ことに気づけず、見出しだけを見ていると機械が引く名前が変わったことに気づけない。

- [x] **依存を Issue 番号で書く形式になっており、説明文でエピック ID を使わないよう
      案内している**
      → `tests/github/issue-template.test.ts`「依存は Issue 番号で書く形式になっている」
      および「依存の説明から『Issue 番号』が消えると違反になる」
      「依存の説明からエピックIDの注意が消えると違反になる」

      テンプレートの `dependencies` の説明文:
      ```
      依存する Issue を **Issue 番号**で書いてください。エピックID だけで書かないで
      ください。両方書く場合は `E2-1（#5）` の形にします。無ければ「なし」。
      ```

- [x] **DoD が唯一の完了定義であることが、テンプレートの説明文に書かれている**
      → `tests/github/issue-template.test.ts`「DoD が唯一の完了定義であることが書かれている」
      と「冒頭の案内にも DoD が唯一の完了定義であることが書かれている」の2本。

      **2箇所に書いた。** フィールドの説明はその欄まで来ないと読まれないので、冒頭の
      案内にも置いた。
      ```
      **これがこの Issue の唯一の完了定義です。** ここに書かれていないことは完了条件に
      なりません。テストに落とせる形で、チェックリストとして書いてください。
      ```

- [x] **テンプレートから実際に Issue を1件作成し、`loop-run` が依存とスコープ外を
      推測せずに読めることを確認した**
      → **#70 を作成した。** テンプレートのフィールド構成そのままで書いてある。

      `loop-run` が読む2箇所を、本文から**推測なしで**引けている:
      ```
      ### 依存

      なし（#37 でテンプレートは入るが、この作業は独立して進められる）
      ```
      ```
      ### スコープ外

      - **エピックIDそのものの廃止** — 表示上の整理として機能しているので残す。…
      - **GitHub のサブイシュー機能への移行** — #37 で検討し、採用しなかった。…
      - **closed Issue の書き換え** — 着手対象にならないので直す価値がない
      - **依存グラフの見直し** — 表記を揃えるだけで、依存関係そのものは変えない
      ```
      従来は `## 依存` の位置も文言も Issue ごとにばらつきがあり、スコープ外の節は
      そもそも無かった（他の Issue を読んで推測していた）。

## mutation test

**`src/` に変更がないため、`CLAUDE.md` の規定では対象外**（同ファイルは
「Issue テンプレート」を除外の例として名指ししている）。ただし**この PR で追加した
検査には壊せる対象がある**ので、テンプレートを壊して検査が落ちることを確認した。

| 壊した内容 | 落ちたテスト |
| --- | --- |
| `id: out-of-scope` を別名（`nonscope`）にする | 2件 |
| 「技術選定の希望・制約」の見出しを別の文言にする | 1件 |
| 依存の説明からエピックIDの注意を消す | 2件 |
| DoD の説明から「唯一の完了定義」を消す | 2件 |
| 「テストで固定すべき境界値」のフィールドを丸ごと削る | 2件 |
| 解析できない字下げ（`  -type: markdown`）にする | 1件 |
| `name:` を消す | 1件 |

**落ちなかった箇所はなし。** 6番目は「解析が壊れたときに違反0件で通らないか」を
見るためのもので、これが落ちることが `type` の検査を入れた理由の裏取りになる。

テンプレートを元に戻して 1160 件すべて通ることを確認済み（38 files / 1160 tests。
1122 → 1160 で +38件）。

## 検証

`npm run check` green（38 files / 1160 tests。修正試行 **0回**）。

**部分解析が本物の YAML と一致していることを別途確認した。** 検査が読む値
（`name` / `description` / 各項目の `type` / `id` / `attributes` の全キー / `required`）を、
コンテナの `python3` + PyYAML で解析した結果と1つずつ比べ、**全項目一致**した。
ブロックスカラー（`|`）と `validations.required` を含む。

```
items: 8 8
all fields match real YAML: True
```

これを確認しないと「自作の部分解析が読めているつもりで別の値を読んでいる」状態が
検査ごと成立してしまう。

## 境界値の確認（`CLAUDE.md` のチェックリスト）

| 境界 | 対応 |
| --- | --- |
| 0件・空 | 空文字のテンプレート / `body:` が空 / フィールドを1つずつ削る（7件） |
| 同時刻 | 該当なし（テンプレートの検査に時刻の概念がない） |
| 日跨ぎ | 該当なし |
| 上限値・範囲外 | `validations` を書かないフィールド、コメント行だけの入力、引用符付きの値 |
| 終端のないデータ | **`validations` の `required` を `attributes` に混ぜない**（`attributes` と同じ 6 スペースで並ぶため、帰属を見ていないと別のキーを読む） |

## #37 の「検討が必要な点」への回答

Issue 本文が判断に含めたいとしていた2点。**どちらも DoD から決まるので、推測では
決めていない。**

- **既存 Issue の移行** → **この PR ではやらない。** #37 の DoD に移行は含まれず、
  Issue 本文も「書き換えるなら別 Issue に分けたほうがよい」としている。
  **#70 として起票した**（対象は #10 / #11 / #20 / #24 の4件。open Issue 全件の
  `## 依存` を確認して数えた）
- **GitHub のサブイシュー機能を使うか** → **使わない。** #37 の DoD が
  「依存を **Issue 番号で書く形式**になっており」を求めており、本文のテキストとして
  読める形が要件になっている。ネイティブの親子関係にすると `loop-run` の読み取り方法を
  変える必要があり、DoD とも噛み合わない

## スコープに入れなかったもの

- **`SKILL.md` からエピックIDの突き合わせ手順を削る** — #37 のスコープは「削れるか
  **検討する**」まで。**削れない**というのが結論で、理由は既存 open Issue に
  エピックIDだけの表記が4件残っていること。**削るのは #70 の担当**（同 Issue の DoD に
  入れてある）
- **既存 Issue の書き換え** — 上記のとおり #70
- **優先度・種別を選ばせる欄** — 優先度はタイトルの `[P1]` が正で、本文にも置くと
  2つの正がぶつかる。#37 のフィールド案にも入っていない

## スタック

- base: `main`（スタックなし。#37 は依存なし）
- 並行して open な PR #68（#58）/ PR #69（#25）とはファイルが重ならない

## 依存の追加

なし。YAML パーサを入れずに済ませた理由は上記「設計の理由」のとおり。

---
_Generated by [Claude Code](https://claude.ai/code/session_01TSp7tp4beNzBvdZGhAJXya)_

### CODE REVIEW

File: .github/ISSUE_TEMPLATE/backlog-item.yml [hunk 1]
```diff
@@ -0,0 +1,93 @@
+name: バックログ項目
+description: tock のバックログに積む作業を1件登録する
+title: "[E?-?][P?] "
+body:
+  - type: markdown
+    attributes:
+      value: |
+        このテンプレートは、**ループ（自動実行）が本文から必要な情報を推測せずに
+        読み取れる形**に揃えるためのものです。空欄にすると、実装側が他の Issue を
```
[the remark this answers is not in the document]
Comment (okrago10) [reply .github/ISSUE_TEMPLATE/backlog-item.yml:9]: 対応しました。**この PR で SKILL の見出しを `##` / `###` 両対応にする**方を選びました（`0f1fb39`）。

再現: `main` の SKILL.md 2-1 を確認したところ、指摘のとおり `## 依存` だけを読む指定でした。

```
### 2-1. 候補の抽出
...
2. 各 Issue の本文にある `## 依存` セクションを読む。…
```

**しかも私自身が #70 を `### 依存` で手書きしています。** Forms がそう展開するからそう書いた
のに、読み取り手順の側を直していませんでした。「テンプレートを入れるだけでは目的を
達成しない」という指摘のとおりです。

**#70 の DoD に足す案を採らなかった理由:** #70 は「既存 Issue の表記を揃える」Issue で、
テンプレートから作った**新規** Issue が読めるかどうかは #37 の成果物の問題です。
#70 のマージを待つ間、テンプレートだけが入って新規 Issue が読めない期間が生まれるのも
指摘のとおりなので、同じ PR で閉じました。

**SKILL.md 2-1:**
```markdown
2. 各 Issue の本文にある依存のセクション（`## 依存` または `### 依存`）を読む。
   **見出しの深さは2種類ある。** 手で書いた Issue は `## 依存`、
   `.github/ISSUE_TEMPLATE/backlog-item.yml`（Issue Forms）から作った Issue は
   `### 依存` になる——Forms は各フィールドを `### <ラベル>` として展開するため。
   同じことが「スコープ外」「受け入れ条件 (DoD)」などの他のフィールドにも当てはまる。
```

**テストで両者を突き合わせました**（`tests/github/issue-template.test.ts` の
「SKILL.md がこのテンプレートの見出しを読める（DoD）」3件）。とくに
「SKILL.md が読む見出しがテンプレートの label と一致している」は、**テンプレートの
`label: 依存` を変えたら落ちます**（変異で確認済み）。片方だけ直して食い違うのを防ぎます。

### 変異で見つけた自分の穴

最初に書いたテストは「SKILL.md のどこかに `### 依存` がある」だけを見ていたため、
**指示を `## 依存` だけに戻しても通ってしまいました**（後段の説明文に `### 依存` が
出るため）。指示そのものの並記（`` `## 依存` または `### 依存` ``）を見る形に直し、
同じ変異で落ちることを確認しました。

| 壊した内容 | 落ちたテスト |
| --- | --- |
| SKILL.md の指示を `##` だけに戻す | **0件 → テストを直して 1件** |
| Forms の展開の説明を消す | 1件 |
| テンプレートの `label: 依存` を変える | 1件 |

`npm run check` green（38 files / 1163 tests）。

**注意:** SKILL.md は PR #72 も触っています（あちらは 2-3 / 3-1 / 「停止するとき」、
こちらは 2-1）。節が違うので機械的なコンフリクトは起きない見込みですが、
先に入った側を取り込むときに確認します。


---
_Generated by [Claude Code](https://claude.ai/code)_

File: tests/github/issue-template.test.ts [hunk 2]
```diff
@@ -0,0 +1,317 @@
...
+
+  // 違反 0 件だけを見ると、解析が実ファイルのフィールドを1件も拾わなくなっても通る
+  it("実テンプレートからフィールドを1件以上拾っている", async () => {
+    const forms = await parseTemplates();
+    const ids = forms.flatMap((form) => form.items.map((item) => item.id));
+
+    expect(ids.filter((id) => id !== undefined).length).toBeGreaterThan(0);
+  });
+
+  it("すべてのテンプレートが方針に従っている", async () => {
```
[the remark this answers is not in the document]
Comment (okrago10) [reply tests/github/issue-template.test.ts]: 対応しました（`0f1fb39`）。指摘のとおり、**自分で書いた `blank_issues_enabled: true` の
理由（枠に収まらない Issue を殺さない）と検査の範囲が矛盾していました。**

再現: 修正前の `templateFiles()` は `config.yml` 以外の全 `.yml` を返し、
`findTemplateViolations` を全件に適用していました。不具合速報用の
`bug-report.yml` を足せば「`id: purpose` のフィールドがありません」で
`npm run check` が落ちます。

提案どおり **`name: バックログ項目` で限定**しました（ファイル名ではなく `name` にしたのは、
ファイルを改名しても検査が外れないようにするため）。

```ts
/** バックログ項目のテンプレート。無ければテストを失敗させる。 */
async function backlogForm(): Promise<IssueForm> {
  const found = (await parseTemplates()).filter((form) => form.name === BACKLOG_NAME);

  expect(found, `name: ${BACKLOG_NAME} のテンプレートが1本だけ見つかること`).toHaveLength(1);

  return found[0] ?? { name: undefined, description: undefined, items: [] };
}
```

**「1本以上ある」を「ちょうど1本ある」に変えました。** 限定する以上、対象が見つからない
（改名・削除された）ことと、2本に増えたことの両方を落とす必要があります。限定しただけだと
「対象が0本なので違反0件」で通ってしまいます。

`npm run check` green（38 files / 1163 tests）。


---
_Generated by [Claude Code](https://claude.ai/code)_
